Native Mode - interface not working / showing in Tcpdump

<!--
Please redirect any questions about Telegraf usage to the InfluxData Community
site: https://community.influxdata.com

Check the documentation for the related plugin including the troubleshooting
section if available.
-->

### Relevant telegraf.conf:
<!-- Place config in the toml code section. -->
```toml
[[inputs.ping]]
        #hostname = "wan_xplornet2"
        method = "native"
        urls = ["8.8.8.8"]
        count = 4
        ping_interval = 1.0
       #interface = "eth1"
[inputs.ping.tags]
interface = "eth1"

[[inputs.ping]]
        #hostname = "wan_xplornet1"
        method = "native"
        urls = ["8.8.4.4"]
        count = 4
        ping_interval = 1.0
        # interface = "eth0"
[inputs.ping.tags]
interface = "eth0"

```

### System info:

<!-- Include Telegraf version, operating system, and other relevant details -->
Telegraf 1.17.1 (git: HEAD 8907e61a)
Distributor ID: Ubuntu
Description:    Ubuntu 20.04 LTS
Release:        20.04
Codename:       focal
Hyper-V VM with 3 nics
X-Influxdb-Version: 1.8.3

### Docker

<!-- If your bug involves third party dependencies or services, it can be very helpful to provide a Dockerfile or docker-compose.yml that reproduces the environment you're testing against -->

### Steps to reproduce:

<!-- Describe the steps to reproduce the bug. -->

1. ... sudo tcpdump -nni eth1 icmp shows no output. 
2.   sudo tcpdump -nni eth0  icmp shows all output
3. ... ping -I eth1 8.8.8.8 works as intended 


### Expected behavior:
See traffic on tcpdump -nni eth1 icmp  and tcpdump -nni eth0  icmp 
<!-- Describe what you expected to happen when you performed the above steps. -->

### Actual behavior:
See all traffic on tcpdump -nni eth0  icmp 
<!-- Describe what actually happened when you performed the above steps. -->
All traffic is going out the default interface and not taking the -I variable (interface = foo)

### Additional info:
> ping,host=c-nms,interface=eth1,url=8.8.8.8 average_response_ms=55.011149,maximum_response_ms=55.011149,minimum_response_ms=55.011149,packets_received=1i,packets_transmitted=1i,percent_packet_loss=0,result_code=0i,standard_deviation_ms=0,ttl=115i 1611812299000000000
> ping,host=c-nms,interface=eth0,url=8.8.4.4 average_response_ms=55.135049,maximum_response_ms=55.135049,minimum_response_ms=55.135049,packets_received=1i,packets_transmitted=1i,percent_packet_loss=0,result_code=0i,standard_deviation_ms=0,ttl=115i 1611812299000000000

root@c-nms:/etc/netplan# cat 50-cloud-init.yaml
network:
    version: 2
    renderer: networkd
    ethernets:
        eth0:
            dhcp4: true
            dhcp4-overrides:
                    route-metric: 100
        eth1:
            dhcp4: true
            dhcp4-overrides:
                    route-metric: 200
        eth2:
            dhcp4: true
            dhcp4-overrides:
                    route-metric: 300

If change to exec mode I get a panic:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b2542a]

goroutine 33 [running]:
github.com/influxdata/telegraf/plugins/inputs/ping.(*Ping).pingToURL(0xc000050460, 0xc000442ee1, 0x7, 0x34ae180, 0xc00021a1c0)
        /go/src/github.com/influxdata/telegraf/plugins/inputs/ping/ping_notwindows.go:22 +0x18a
github.com/influxdata/telegraf/plugins/inputs/ping.(*Ping).Gather.func1(0xc000050460, 0x34ae180, 0xc00021a1c0, 0xc000442ee1, 0x7)
        /go/src/github.com/influxdata/telegraf/plugins/inputs/ping/ping.go:138 +0x9b
created by github.com/influxdata/telegraf/plugins/inputs/ping.(*Ping).Gather
        /go/src/github.com/influxdata/telegraf/plugins/inputs/ping/ping.go:131 +0xd2
<!-- Include gist of relevant config, logs, etc. -->

--- a/plugins/inputs/ping/ping.go
+++ b/plugins/inputs/ping/ping.go
@@ -4,6 +4,7 @@ import (
 	"errors"
 	"fmt"
 	"math"
+	"net"
 	"os/exec"
 	"runtime"
 	"strings"
@@ -28,6 +29,8 @@ type Ping struct {
 	// Pre-calculated interval and timeout
 	calcInterval time.Duration
 	calcTimeout  time.Duration
+
+	sourceAddress string
 
 	Log telegraf.Logger `toml:"-"`
 
@@ -171,6 +174,7 @@ func (p *Ping) nativePing(destination string) (*pingStats, error) {
 		pinger.SetNetwork("ip6")
 	}
 
+	pinger.Source = p.sourceAddress
 	pinger.Interval = p.calcInterval
 	pinger.Timeout = p.calcTimeout
 
@@ -308,6 +312,23 @@ func (p *Ping) Init() error {
 		p.calcTimeout = time.Duration(5) * time.Second
 	} else {
 		p.calcTimeout = time.Duration(p.Timeout) * time.Second
+	}
+
+	// Support either an IP address or interface name
+	if p.Interface != "" {
+		if addr := net.ParseIP(p.Interface); addr != nil {
+			p.sourceAddress = p.Interface
+		} else {
+			i, err := net.InterfaceByName(p.Interface)
+			if err != nil {
+				return fmt.Errorf("Failed to get interface: %w", err)
+			}
+			addrs, err := i.Addrs()
+			if err != nil {
+				return fmt.Errorf("Failed to get the address of interface: %w", err)
+			}
+			p.sourceAddress = addrs[0].(*net.IPNet).IP.String()
+		}
 	}
 
 	return nil

